Step4 - 자동차 경주(우승자)

catsbi opened the pull request:
안녕하세요 리뷰어님. 매번 꼼꼼한 리뷰덕분에 너무 많이 배우는 것 같네요 👍 
이번 과정 끝나고 7월에 회사 팀내에서도 코틀린 스터디를 진행하고 4분기에는 프로젝트 모듈 하나씩 코틀린으로 넘어가려고 하는데, 많은 도움이 됩니다 👏

이전 미션에서 주신 피드백 반영과, 자동차 이름 추가 및 우승자 출력 로직까지 작성했습니다. 


Q1. 이동에 대해 후진까지 고려를 해서 리팩토링 하다보니까, 오버엔지니어링이 아닐까 하는 고민이 들더라구요. 
리뷰어님은 기능 요구사항에 대해서 어디까지 확장성을 고려하시는 편인가요? 

Q2.

catsbi committed:
refactor(racing): 피드백 반영 리팩토링

- infix method 작성 방식 수정
- Car equal 재정의 메서드 로직 변경
- Cars 일급 컬렉션 델리게이션 사용
- 이동 로직 추상화
- UI Input 로직 꼬리재귀로 리팩토링
- UI Output 로직 리팩토링

catsbi committed:
refactor(racing): 피드백 반영 리팩토링

- UI Input 로직 꼬리재귀로 리팩토링

catsbi committed:
refactor(MoveStrategy): 람다 매개변수 위치 변경

- 람다 매개변수를 끝으로 위치 변경

catsbi committed:
feat(racing): 자동차 경주 우승자 기능 구현

- 자동차 이름 프로퍼티 추가및 관련 값 객체(Name) 구현
- 경주 우승자 판별 로직 구현
- (RacingResult)경주 우승자 포함 데이터 클래스 구현
- 자동자 이름 중복 체크 로직 구현

saintbeller96 commented on src/main/kotlin/racing/ui/RacingController.kt:
```diff
@@ -11,12 +12,12 @@ class RacingController(
 
     fun startRacing() {
-        val carCapacity = racingInput.requestNumberOfCars()
+        val cars = Cars(racingInput.requestCarNames())
```
컨트롤러 계층에서 `Cars` 도메인를 생성하는 게 조금 어색해보여요.
`RaceService`에서 `Cars`를 생성하는 건 어떻게 생각하시나요??

catsbi replied:
@saintbeller96 
이 부분이 어색한 이유가 스프링과 같은 웹 프레임워크에서는 UI -> Controller -> Serivce 플로우에서 유효성 검증을 책임질 Validator 계층이 없고 ArgumentResolver도 없다보니 해당 책임을 컨트롤러가 가지게 되고, 유효성 검증 로직을 도메인의 생성자에 의존하다보니 이렇게 된 것 같습니다. 

근데 생각해보니 Validator가 없더라도 비즈니스용 프로퍼티 유효성 검증에 대해서는 Request 객체에도 존재했고 
그렇기에 Save, Update 각각 다른 유효성 검증 로직을 가질 수 있다는게 생각이나서 RacingGameRequest 객체를 만들고 내부에 유효성 검증 로직을 만들었습니다. 이는 스프링의 Bean Validation을 참고했습니다.

saintbeller96 commented on src/main/kotlin/racing/ui/RacingInput.kt:
```diff
@@ -1,26 +1,31 @@
-        } catch (e: NumberFormatException) {
-            println("숫자를 입력해주세요.")
-            requestNumberOfCars()
+    tailrec fun requestCarNames(): String {
```
`tailrec` 활용👍👍👍

saintbeller96 commented on src/main/kotlin/racing/ui/RacingInput.kt:
```diff
@@ -1,26 +1,31 @@
+        return if (carNames != null) carNames else {
+            println("유효한 자동차 이름을 입력하세요. Input: $carNames")
+            requestCarNames()
         }
```
[코틀린 코드 컨벤션](https://developer.android.com/kotlin/style-guide?hl=ko#formatting)에 따르면 `if .. else ..` 표현식을 한줄로 표현하지 못하는 경우, 모든 body엔 중괄호를 사용하는 것을 권장합니다🙂

saintbeller96 commented on src/main/kotlin/racing/application/RacingService.kt:
```diff
@@ -1,17 +1,29 @@
+    fun racing(racingInfo: RacingInfo): RacingResult {
 
-        return racingInfo.round.map { currentRound ->
-            players.notifyMoving(moveStrategy = moveStrategy)
-            RacingHistory(round = currentRound, records = players.toMap({ it.getId }, { it.getDistance }))
+        val racingHistories = racingInfo.round.map { currentRound ->
+            racingInfo.cars.notifyMoving(movableStrategy = movableStrategy)
+            RacingHistory(round = currentRound, records = racingInfo.cars.associateBy({ it.name }, { it.distance }))
         }
+
+        return RacingResult(
+            racingHistories = racingHistories,
+            winners = findWinners(racingHistories)
+        )
+    }
```
`RaceInfo` 내부에 있는 `round`와 `cars`에 많은 메시지를 던지고 있는 것 같아요.

`RaceInfo`가 단순히 데이터만 전달해주는 것이 아닌, 경주를 진행하고 기록을 제공할 수 있는 도메인 객체로 만들어보시면 어떨까요?

catsbi replied:
RacingInfo대신 RacingGameRequest 객체로 바뀌게되면서 도메인 객체(Cars, round, ...)의 생성이 Service에서 이뤄져 지역변수로 선언되면서 request 객체는 순수하게 자료 전달용으로 변경하였습니다.

saintbeller96 commented on src/main/kotlin/racing/application/RacingService.kt:
```diff
@@ -1,17 +1,29 @@
+        val lastHistory = racingHistories.last().records
+        val winnerDistance = lastHistory.maxBy { it.value }.value
+
+        return lastHistory.filter { winnerDistance == it.value }.map { it.key }
```
체이닝 되는 함수는 줄바꿈으로 표현해주시면 좋을 것 같아요! [관련 컨벤션](https://kotlinlang.org/docs/coding-conventions.html#wrap-chained-calls)

catsbi replied:
컨벤션은 읽어봤는데, 고민이 되는 부분이, 보통 개행에 관련해서는 디미터 법칙을 빼놓고 말할수가 없는데 디미터 법칙에서는 객체 엔티티 그래프 탐색에 있어서 이웃의 이웃을 알게되는 경우가 없게끔 하라고 하면서 구조를 바꾸거나 응답 결과가 호출자와 무관한 경우 개행을 하라고 하는데, 자바에서는 Stream API가 이런 비슷한 이유로 논의가 되는데, Stream API는 반환값도 Stream 객체이기 때문에 디미터법칙에 해당되지 않고 가독성이라는 이유를 제외하고는 개행을 강제할 필요가 없다고, 하고 있는걸로 알아요. 

코틀린에서는 어떤 이유로 개행을 해야 하는지 컨벤션을 봐서는 잘 모르겠네요 ㅠㅠ.. 

(논외로 자바에서도 가독성이나 개념적 분리를 위해 개행을 하고 있기는 합니다. )

saintbeller96 replied:
저도 체이닝 함수를 개행하는 건 가독성이 가장 크다고 생각해요.
코드 스타일은 강제할 순 없지만, 코틀린은 공식 문서에서 다양한 경우에 대한 컨벤션을 제공하고 있으니 만큼 참고해주시면 좋을 것 같습니다!

saintbeller96 commented on src/main/kotlin/racing/domain/Capacity.kt:
```diff
@@ -15,7 +15,7 @@ value class Capacity(val value: Int) : Iterable<Int>, Comparable<Capacity> {
 
     override fun iterator(): Iterator<Int> = (0 until value).iterator()
```
이전 [comment](https://github.com/next-step/kotlin-racingcar/pull/1071#discussion_r1224546405)에서 답변 주신 부분이네요🙂
`Capacity` 객체가 IntRange를 제공해야 한다면 프로퍼티로 제공해주면 어떨까요?
```suggestion
   val availableRange: IntRange get() = 0 until value
```

catsbi replied:
- 프로퍼티로 제공할지 fun 함수로 제공할지 어떻게 결정해야 할지 아직 헷갈리는데, 어떤 지침이나 규칙이 있을까요? 
- Interable 을 구현해서 해당 객체가 반복가능하다고 알려주는 것 대신 IntRange를 프로퍼티로 제공할 때 어떠한 강점이 있을까요? 
- 우선 해당 내용은 반영했습니다.

saintbeller96 replied:
1. [코틀린 공식 문서](https://kotlinlang.org/docs/coding-conventions.html#functions-vs-properties)를 참고하시면 좋을 것 같아요.

2. `수량, 용량`은 일반적으로 반복이라는 속성이 없는 개념인 만큼 `Iterator`의 하위 타입이 되는 게 어색하다고 생각했어요. 이는 `Capacity`를 사용하는 사용자들에게 혼란을 줄 수도 있다고 생각합니다(저도 그랬구요😅).
대신 `수량만큼의 범위 제공`이라는 기능을 한정적으로 제공함으로써 사용자들이 제공자의 의도를 명확하게 파악할 수 있다는 장점이 있을 것 같아요. 마치 `Composition`을 활용한 것처럼 말이죠🙂

saintbeller96 commented on src/main/kotlin/racing/domain/Cars.kt:
```diff
@@ -1,22 +1,35 @@
-    override fun iterator(): Iterator<Car> = values.iterator()
+        private fun createCarList(names: String): Set<Car> {
+            val splitNames = names.split(DELIMITER)
+            val result = splitNames.distinct()
+                .map { it.trim() }
+                .mapIndexed { idx, name ->
+                Car(id = idx, name = Name(name))
+            }.toSet()
 
-    companion object {
-        private fun createCarList(capacity: Capacity): List<Car> {
+            require( splitNames.size == result.size ) {
+                "자동차 이름은 중복될 수 없습니다."
+            }
+
+            return result
+        }
```
String 인자를 받아 split한 후 Cars를 생성하는 과정이 Cars에 있는게 조금 어색해보이는 것 같아요👀
String이 아닌 `List<Name>`이나 `Set<Name>`을 받아 Cars를 생성해주는 방법은 어떻게 생각하시나요?

catsbi replied:
Input Layer에서 최대한 값 조작을 없도록 하고싶은게 컸습니다. 
생성자에서 문자열을 자르고 만드는건 좀 과할 수 있을 것 같긴하네요. 
그런데 정적 팩토리 메서드를 사용하면 이 부분은 괜찮을 것 같습니다. 

다만 Web환경에서 requestParam에도 "a,b,c,d,e"로 value를 전달하면 List로 파싱되서 들어가는걸 생각해보면 Input -> mapping request -> controller 에서 가운데 mapping request과정에서 String to List변환이 되면 될 것 같아 RacingGameRequest에 반영했습니다!

saintbeller96 commented on src/main/kotlin/racing/domain/Cars.kt:
```diff
@@ -1,22 +1,35 @@
             require(capacity > Capacity.ZERO) { "Invalid Size: ${capacity.value} required 0 over size" }
 
-            return capacity.map { Car(id = it) }
+            return capacity.map { Car(id = it) }.toSet()
```
`Capacity`가 `Iterator`의 하위 타입이기 때문에 `map`과 같은 다양한 연산을 수행할 수 있군요!

하지만 `Capacity`가 수량, 용량이라는 개념을 표현한다면, 내부를 순회하며 mapping 하는 기능이 `Capacity`에 있는 건 조금 어색한 것 같아요.

`repeat`이나 `buildSet { }` 같은 함수를 통해 `Capacity`의 크기만큼 Cars를 만든다는 과정을 조금 더 명확하게 표현해보면 어떨까요??

catsbi replied:
```kotlin
return buildSet {
    repeat(capacity.value) {
        add(Car(id = it))
    }
}
```

이런식 말일까요? 나쁘지 않은데 아까 위에서 제안해주신 availableRange 나아보여서 해당 방법으로 구현했습니다!

saintbeller96 commented on src/main/kotlin/racing/domain/strategy/MovableStrategy.kt:
```diff
@@ -0,0 +1,7 @@
+import racing.domain.DirectionType
+
+fun interface MovableStrategy {
+    fun movable(): DirectionType
```
사소한 부분이긴 하지만 `movable`이라는 이름만 보았을 땐`Boolean` 타입을 반환할 것 같다는 생각이 들었어요.
이동 방향을 제공한다는 의미라면 `navigate` 같은 함수명은 어떨까요??

catsbi replied:
공감합니다. 계속 추가되다보니 나중엔 여유가 없어서 좀 신경을 덜 쓴것 같네요 😥

saintbeller96 commented on src/main/kotlin/racing/domain/strategy/StopOrStraightMoveStrategy.kt:
```diff
@@ -0,0 +1,13 @@
+
+class StopOrStraightMoveStrategy(private val additionalDistance: Distance) : MoveStrategy {
+
+    override fun move(baseDistance: Distance, movableStrategy: MovableStrategy): Distance {
+        return totalDistance(baseDistance).takeIf { movableStrategy.movable() == DirectionType.STRAIGHT } ?: baseDistance
+    }
```
이전 단계에서 `Car`의 `distance`에 새로운 `Distance`를 더해주는 방식도 좋다고 생각했어요👍
변경하신 부분에서 `baseDistance`를 인자로 받으신 의도가 궁금합니다!

catsbi replied:
전진, 정지, 후진만 고려하면 Distance만 고려하면되서 값에 대한 증가/유지/감소만 신경쓰면되었는데, 
좌회전이나 우회전같은 타입까지 고려하다보니 더해주는 방식 하나로는 대응이 힘들 것 같았습니다. 그리고  Distance는 0이상의 숫자만을 취급하기에 이에 대한 체킹도 모두 Car에 넣는것보단 이동 전략 구현체에서 책임지는게 맞을 것 같아 옮기게 되었습니다.

saintbeller96 commented on src/test/kotlin/racing/application/RacingServiceTest.kt:
```diff
@@ -1,27 +1,32 @@
+            val racingService = RacingService { DirectionType.STRAIGHT }
 
-            and("유효한 횟수와 자동차 대수를 인자로 전달할 경우 ") {
-                val info = RacingInfo(round = Capacity(5), capacity = Capacity(5))
+            And("유효한 횟수와 자동차 참가자를 인자로 전달할 경우 ") {
+                val info = RacingInfo(round = Capacity(5), cars = Cars("a,b,c,d,e"))
                 val actual = racingService.racing(info)
 
-                then("전달받은 자동차 대수만큼 자동차를 생성하고 결과를 반환한다.") {
-                    actual shouldHaveSize 5
-                    actual.mapIndexed { idx, info ->
+                Then("전달받은 자동차 대수만큼 자동차를 생성하고 결과를 반환한다.") {
+                    actual.racingHistories shouldHaveSize 5
+                    actual.racingHistories.mapIndexed { idx, info ->
                         info.records.forEach {
                             it.value shouldBe Distance(idx + 1L)
                         }
                     }
+                    actual.winners shouldBe arrayOf(Name("a"),Name("b"),Name("c"),Name("d"),Name("e"))
                 }
             }
```
BDD 방식으로 테스트를 잘 작성해주신 것 같아요👍👍👍

저는 BDD 방식을 사용할 땐 [Cucumber에서 제공하는 예시](https://cucumber.io/docs/bdd/)들을 많이 참고하는 편이에요.

한솔님도 잘 아시고 계시겠지만 BDD에서 주로 사용하는 Given/When/Then은 일반적으로 다음과 같이 표현할 수 있는데요,
- Given: 주어진 조건, 상황
- When: 발생한 사건, 행위
- Then: 그로 인한 결과 

이를 바탕으로 위 코드를 다음과 같이 표현할 수도 있을 것 같아요. 

물론 테스트를 작성하는 스타일 차이지만..! 이런 방식도 있다는 정도로 생각해주시면 좋을 것 같습니다🙂
```suggestion
    Given("유효한 횟수와 자동차 참가자가 있다") {
        val info = RacingInfo(round = Capacity(5), cars = Cars("a,b,c,d,e"))
        And("전진 전략이 항상 전진이다") {
            val racingService = RacingService { DirectionType.STRAIGHT }
            When("경주를 진행하면") {
                val actual = racingService.racing(info)
                Then("경주 결과를 확인할 수 있다") {
                    // 검증
                }
            }
        }
    }
```

catsbi replied:
given절에 계속 when 내용이 들어가는 것 같아 고민을 좀 하고 있었는데 리뷰어님이 제안해주신 부분이 좀 더 설득력이 있는 내용같습니다 ㅎㅎ

saintbeller96 commented on src/test/kotlin/racing/domain/strategy/StopOrStraightMoveStrategyTest.kt:
```diff
@@ -0,0 +1,36 @@
+import racing.domain.DirectionType.STOP
+import racing.domain.Distance
+
+class StopOrStraightMoveStrategyTest : BehaviorSpec({
+    Given("전진 혹은 멈춤 전략은") {
+        When("이동 속도가 1이고") {
+            val strategy = StopOrStraightMoveStrategy(Distance(1L))
+
+            And("이동 타입이 STRAIGHT 인 경우") {
+                val actual = strategy.move(movableStrategy = { DirectionType.STRAIGHT }, baseDistance = Distance.ZERO)
+
+                Then("이동 거리는 1만큼 증가한다") {
+                    actual shouldBe Distance(1L)
+                }
+            }
+
+            listOf(STOP, BACK, LEFT, RIGHT).forEach {
+                And("이동 타입이 $it 인 경우") {
+                    val actual = strategy.move(movableStrategy = { it }, baseDistance = Distance.ZERO)
+
+                    Then("멈춰 있는다.") {
+                        actual shouldBe Distance.ZERO
+                    }
+                }
+            }
+        }
+    }
+})
```
Kotest는 BDD의 또 다른 방식인 DescribeSpec을 지원합니다!

DescribeSpec은 BehaviorSpec과는 달리 아래 패턴으로 구성되는데요, 
- Describe: 묘사할 대상
- Context: 주위의 환경, 상황
- It: 대상의 행동 

개인적으로 어떤 객체의 행위를 묘사하고자 할때 적합한 방식이라고 생각해요.
(Javascript 진영에서 이런 스타일을 사용한다고 하더라구요)

이를 바탕으로 `이동 전략`을 테스트하면 다음과 같이 나타낼 수 있을 것 같아요.
그냥 이런 방식도 있구나 정도로 생각해주시면 좋을 것 같습니다!
```suggestion
class StopOrStraightMoveStrategyTest : DescribeSpec({
    describe("전진 혹은 멈춤 전략은") {
        context("이동 속도가 1이고") {
            val strategy = StopOrStraightMoveStrategy(Distance(1L))

            context("이동 타입이 STRAIGHT 인 경우") {
                val actual = strategy.move(movableStrategy = { DirectionType.STRAIGHT }, baseDistance = Distance.ZERO)

                it("이동 거리를 1만큼 증가시킨다") {
                    actual shouldBe Distance(1L)
                }
            }
        }
    }
})
```

saintbeller96 reviewed (changes requested):
안녕하세요 한솔님!
자동차 경주(우승자) 구현도 너무 잘 해주신 것 같아요💯
몇가지 피드백을 남겼는데 확인해보시고 다시 리뷰 요청 부탁드려요🙂

saintbeller96 commented:
> 안녕하세요 리뷰어님. 매번 꼼꼼한 리뷰덕분에 너무 많이 배우는 것 같네요 👍 이번 과정 끝나고 7월에 회사 팀내에서도 코틀린 스터디를 진행하고 4분기에는 프로젝트 모듈 하나씩 코틀린으로 넘어가려고 하는데, 많은 도움이 됩니다 👏
> 
> 이전 미션에서 주신 피드백 반영과, 자동차 이름 추가 및 우승자 출력 로직까지 작성했습니다.
> 
> Q1. 이동에 대해 후진까지 고려를 해서 리팩토링 하다보니까, 오버엔지니어링이 아닐까 하는 고민이 들더라구요. 리뷰어님은 기능 요구사항에 대해서 어디까지 확장성을 고려하시는 편인가요?
> 
> Q2. 

저는 보통 **주어진 요구사항**을 기반으로 **도메인으로 표현하고자 하는 현실 개체**를 보며 **일반적으로 변경될 수 있는 부분**에 대해선 확장을 고려해 설계하려고 노력하고 있어요.

말이 너무 모호한 것 같아서 예시를 들어보겠습니다😅

자동차 경주 미션의 요구사항 중 하나로 자동차는 이동, 정지 두가지 기능을 수행할 수 있어야 합니다. 

하지만 자동차라는 현실세계의 개체는 이동, 전진 뿐만 아니라 다양한 방향으로 이동할 수 있죠!

도메인 모델과 현실 개체가 100% 일치하는 것은 아니지만, 이렇게 표현하고자 하는 현실 개체를 보면서 어느 부분에서 변경될 가능성이 있겠구나 정도를 생각해보는 것 같아요.

catsbi committed:
refactor(CarController): 피드백 반영

- 컨트롤러의 Cars 도메인 생성 로직 서비스 계층으로 분리
- RacingInput -> RacingGameRequest로 변경
- Request 객체에 유효성 검증 로직 추가
- 표현식 if-else 중괄호 추가

catsbi committed:
refactor(CarController): 피드백 반영

- Capacity가 Iterable을 구현하지 않도록 변경
- 이동 전략 인터페이스 네이밍 변경 -> Navigator

catsbi commented:
리뷰어님 피드백 주신부분들 최대한 확인 후 반영해봤습니다. 항상 리뷰 감사합니다 :)

saintbeller96 reviewed (approved):
안녕하세요 한솔님!

자동차 경주(우승자) 구현하시느라 고생 많으셨습니다!

이미 MVC 패턴을 활용해 잘 구현해주셔서 5단계는 추가로 구현이 필요 없을지도 모르겠네요.

혹시 더 리뷰받고 싶은 부분이 있으시면 코멘트로 남겨주세요🙂

충분하다고 생각하시면 다음 단계는 바로 merge 하겠습니다!
